model/flow/sealing_segment.go

```diff
@@ -0,0 +1,149 @@
+	ErrSegmentResultLookup       = fmt.Errorf("failed to lookup execution result")
+)
+
+type SealingSegmentBuilder struct {
+	resultLookup    func(resultID Identifier) (*ExecutionResult, error)
+	includedResults map[Identifier]*ExecutionResult
+	blocks          []*Block
+	results         []*ExecutionResult
+}
+
+// AddBlock appends block to blocks
+func (builder *SealingSegmentBuilder) AddBlock(block *Block) error {
+	//sanity check: block should be 1 height higher than current highest
+	if !builder.isValidHeight(block) {
+		return fmt.Errorf("invalid block height (%x): %w", block.Header.Height, ErrSegmentInvalidBlockHeight)
+	}
+
+	// cache results in included results
+	// they could be referenced in a future block in the segment
+	for _, result := range block.Payload.Results.Lookup() {
+		builder.includedResults[result.ID()] = result
+	}
+
+	for _, receipt := range block.Payload.Receipts {
+		if _, ok := builder.includedResults[receipt.ResultID]; !ok {
+			result, err := builder.resultLookup(receipt.ResultID)
+			if err != nil {
+				return fmt.Errorf("%w: (%x) %v", ErrSegmentResultLookup, receipt.ResultID, err)
+			}
+
+			builder.addExecutionResult(result)
+			builder.includedResults[receipt.ResultID] = result
+		}
+	}
+	builder.blocks = append(builder.blocks, block)
+	return nil
+}
+
+// AddExecutionResult adds result to executionResults
+func (builder *SealingSegmentBuilder) addExecutionResult(result *ExecutionResult) {
+	builder.results = append(builder.results, result)
+}
+
+// SealingSegment will check if the highest block has a seal for the lowest block in the segment
+func (builder *SealingSegmentBuilder) SealingSegment() (*SealingSegment, error) {
+	segment := &SealingSegment{
+		Blocks:           builder.blocks,
+		ExecutionResults: builder.results,
+	}
+
+	if len(segment.Blocks) < 1 {
+		return nil, fmt.Errorf("expect at least 2 blocks in a sealing segment or 1 block in the case of root segments, but actually got %v: %w", len(segment.Blocks), ErrSegmentBlocksWrongLen)
+	}
+
+	// if root sealing segment skip seal sanity check
+	if len(segment.Blocks) == 1 {
+		return segment, nil
+	}
+
+	if !builder.hasValidSeal() {
+		return nil, fmt.Errorf("sealing segment missing seal lowest (%x) highest (%x): %w", segment.Lowest().ID(), segment.Highest().ID(), ErrSegmentMissingSeal)
+	}
+
+	return segment, nil
+}
+
+// isValidHeight returns true block is exactly 1 height higher than the current highest block in the segment
+func (builder *SealingSegmentBuilder) isValidHeight(block *Block) bool {
+	if builder.highest() == nil {
+		return true
+	}
+
+	return block.Header.Height == builder.highest().Header.Height+1
+}
+
+// hasValidSeal returns true if highest block in the segment contains a seal for the lowest block
+func (builder *SealingSegmentBuilder) hasValidSeal() bool {
+	for _, seal := range builder.highest().Payload.Seals {
+		if seal.BlockID == builder.lowest().ID() {
```

Comment:
better to cache the `ID()`, otherwise we will compute over and over

Reply:
https://github.com/onflow/flow-go/pull/1533/commits/00a77e58055c9b7fbc3a8d569aa91866610f5646